Source/Core/Common/MemoryUtil.cpp

```diff
@@ -29,46 +29,109 @@
 #endif
 #endif
 
-// Valgrind doesn't support MAP_32BIT.
-// Uncomment the following line to be able to run Dolphin in Valgrind.
-//#undef MAP_32BIT
-
 namespace Common
 {
-#if !defined(_WIN32) && defined(_M_X86_64) && !defined(MAP_32BIT)
+#if defined(_M_X86_64)
+#if defined(_WIN32)
+static SYSTEM_INFO sys_info;
+#define MEM_PAGE_SIZE (sys_info.dwPageSize)
+#else
 #include <unistd.h>
+#define MEM_PAGE_SIZE (getpagesize())
+#endif
+
 static uintptr_t RoundPage(uintptr_t addr)
 {
-  uintptr_t mask = getpagesize() - 1;
+  uintptr_t mask = MEM_PAGE_SIZE - 1;
   return (addr + mask) & ~mask;
 }
 #endif
 
+static int hint_dummy;
+static const uintptr_t hint_location = reinterpret_cast<uintptr_t>(&hint_dummy);
+
+#if defined(_WIN32) && defined(_M_X64)
+static uintptr_t last_executable_addr;
+static void* SearchForFreeMem(size_t size)
+{
+  if (!last_executable_addr)
+    last_executable_addr = hint_location - sys_info.dwPageSize;
+  last_executable_addr -= size;
+
+  MEMORY_BASIC_INFORMATION info;
+  while (VirtualQuery(reinterpret_cast<void*>(last_executable_addr), &info, sizeof(info)) == sizeof(info))
+  {
+    // went too far, unusable for executable memory
+    if (last_executable_addr + 0x80000000 < hint_location)
+      return nullptr;
+
+    uintptr_t end = last_executable_addr + size;
+    if (info.State != MEM_FREE)
+    {
+      last_executable_addr = reinterpret_cast<uintptr_t>(info.AllocationBase) - size;
+      continue;
+    }
+
+    if (reinterpret_cast<uintptr_t>(info.BaseAddress) + info.RegionSize >= end &&
```

Comment:
You can do

```cpp
const auto base_address_value = reinterpret_cast<uintptr_t>(info.BaseAddress);
if (base_address_value + info.RegionSize >= end && base_address_value <= last_executable_addr)
  return reinterpret_cast<void*>(last_executable_addr);
```

So that you don't need to repeat the cast here.